Add: added filter for picks page

ecila7290 opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)

#### Add: added detail page API
 - 상세페이지를 불러오는 API작성

#### Edit: edited detail page view
- 리턴하는 딕셔너리에 키 이름 추가

#### Edit: edited detail page view
- hotel_id 1을 제외한 인스턴스를 가져오지 못하는 오류 해결

#### Add: added filter for picks page
 - 지역, 카테고리, 가격에 따른 필터 설정
 - 가격 필터는 모든 호텔을 반복문으로 돌면서 각 호텔 인스턴스에 aggregate를 사용하여 호텔의 방 중 최저가를 구한 후, 최저가가 쿼리스트링으로 받은 가격 범위 안에 있다면 호텔 id를 리스트에 넣어줌.
 - 리스트를 filter(id__in=리스트)로 리스트에 넣은 id에 해당하는 인스턴스만 가져옴

#### Edit: edited PicksView's min_price&max_price type
 - 리턴하는 min_price, max_price의 형식을 스트링+콤마에서 정수형으로 수정

#### Add: added number of rooms
 - 프론트에 리턴하는 정보에 방의 개수를 추가
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- id '@' 포함, pw 5 글자 이상 입력 시 버튼 활성화(예시)

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
hotel/PicksView에서 필터링 하는 로직이 너무 긴거같은데 심플하게 할 수 있는 방법이 없을까요

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [ ] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [ ] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [ ] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [ ] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

ecila7290 committed:
add: added detail page API
 - 상세페이지를 불러오는 API작성

ecila7290 committed:
Edit: edited detail page view
 - 리턴하는 딕셔너리에 키 이름 추가

ecila7290 committed:
Edit: edited detail page view
 - hotel_id 1을 제외한 인스턴스를 가져오지 못하는 오류 해결

ecila7290 committed:
Add: added filter for picks page
 - 지역, 카테고리, 가격에 따른 필터 설정
 - 가격 필터는 모든 호텔을 반복문으로 돌면서 각 호텔 인스턴스에 aggregate를 사용하여 호텔의 방 중 최저가를 구한 후, 최저가가 쿼리스트링으로 받은 가격 범위 안에 있다면 호텔 id를 리스트에 넣어줌.
 - 리스트를 filter(id__in=리스트)로 리스트에 넣은 id에 해당하는 인스턴스만 가져옴

ecila7290 committed:
Add: added replace to price to remove commas
 - 쿼리스트링으로 받는 금액에서 콤마를 제거하는 replace 메소드 추가

ecila7290 committed:
Edit: edited picks views and detail page view to use select_related & preftech_related
 - select_related와 prefetch_related로 데이터를 캐싱하여 쿼리횟수를 줄임

ecila7290 committed:
Edit: edited unneccesary print
 - 불필요한 print문 삭제

ecila7290 committed:
Edit: edited checkin/checkout time
 - DetailPageView에서 체크인/체크아웃 시간이 '시:분:초' 였던 것을 '시:분'으로 수정

ecila7290 committed:
Add: added keys in detail page view json
 - DetailPageView에서 리턴하는 json에 방id와 이미지링크 추가

ecila7290 committed:
Edit: edited PicksView's min_price&max_price type
 - 리턴하는 min_price, max_price의 형식을 스트링+콤마에서 정수형으로 수정

ecila7290 committed:
Add: added number of rooms
 - 프론트에 리턴하는 정보에 방의 개수를 추가

ecila7290 committed:
Add: added hotel image url
 - 상세 페이지에서 보여줄 호텔의 메인 이미지 url 추가

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
 
 class MainBannerView(View):
     def get(self, request):
         hotels  = Hotel.objects.all()[:10]
```
한 페이지에 나올 데이터 수가 10개인가요?
이렇게 숫자를 적어두시는 것은 하드코딩입니다. 함수 상단에서 Constant로 지정해주시거나, 아니면 서비스의 확장성을 위해서 프론트엔드에서 숫자를 받아오시는 것도 한 방법입니다. 서비스의 기획이 변경되어 한 페이지에 15장씩 보여주게 되었을 때, 프론트엔드에서 자유롭게 그 수를 변경할 수 있도록 자유도를 주는 것이 조금 더 확정성있기 때문입니다.

ecila7290 replied:
프론트와 상의해서 쿼리스트링으로 받아서 처리하기로 했습니다.

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
-            'desc_title': hotel.description_title,
-            'desc'      : hotel.description_first,
-            'image'     : hotel.thumbnail_url,
+            'name'             : hotel.name,
+            'category'         : hotel.category.name,
+            'location'         : hotel.location.name,
+            'description_title': hotel.description_title,
+            'description_first': hotel.description_first,
+            'thumbnail_url'    : hotel.thumbnail_url,
```
good !

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
+class PicksView(View):
+    def get(self, request):
+        hotels=Hotel.objects.select_related('category','location').prefetch_related('room_set','tags').all().order_by('id')
+        filter_set=[]
+        if request.GET.get('category'):
+            filter_set.append({'category__name':request.GET['category'].strip("'")})
+        if request.GET.get('location'):
+            filter_set.append({'location__name__contains':request.GET['location'].strip("'")})
+
+        for field in filter_set:
+            if filter_set == []:
+                break
+            hotels=hotels.filter(**field)
```
필터 셋을 설정하시는 것 좋은 아이디어입니다. 그러나 이 필터들이 리스트로 들어올 경우 순회를 해야하고, 혹시 필터의 갯수가 늘어나게 된다면 그만큼 for loop을 더 돌아야 되는 효율의 문제가 생깁니다.
따라서 이 필터를 딕셔너리로 한 번 바꿔보시면 더 좋을 것 같습니다 :)

ecila7290 replied:
딕셔너리로 바꿔봤는데 맞는지 확인 부탁드립니다!

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
+                break
+            hotels=hotels.filter(**field)
+
+        if request.GET.get('price'):
+            price=request.GET['price'].strip("'").split('~')
+            if price[0] == '':
+                min_price = 0
+                max_price = int(price[1].replace(",","").strip('원'))
+            elif price[1] == '':
+                min_price = int(price[0].replace(",","").strip('원'))
+                max_price = int(1e8)
+            else:
+                min_price = int(price[0].replace(",",""))
+                max_price = int(price[1].replace(",","").strip('원'))
```
이 과정을 백엔드에서 모두 진행한 뒤 보내주기로 약속하신건가요?

ecila7290 replied:
처음에 페이지를 로드할 때 가격 범위('100,000~200,000원' 과 같은 형식)를 백엔드의 데이터베이스에서 보내주고, 가격 기준으로 필터링을 하게 되면 쿼리스트링에 백엔드에서 보냈던 범위 형식을 그대로 받기 때문에 콤마와 원을 백에서 처리하게 되었습니다.

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
+            else:
+                min_price = int(price[0].replace(",",""))
+                max_price = int(price[1].replace(",","").strip('원'))
+            hotel_price_range=[hotel.id for hotel in hotels if min_price<= hotel.room_set.aggregate(p=Min('price'))['p']<max_price]
```
aggregate good!

ecila7290 replied:
👍

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
+
+            hotels=hotels.filter(id__in=hotel_price_range)
+
+        offset = int(request.GET.get('offset'))
+        LIMIT  = int(request.GET.get('limit'))
```
`LIMIT`은 왜 대문자인가요?

ecila7290 replied:
보내는 데이터의 limit은 고정되어 있다고 생각해서(constant) 대문자로 설정했습니다.

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
+
+class DetailPageView(View):
+    def get(self, request,hotel_id):
+        rooms=Room.objects.select_related('hotel').prefetch_related('roomimage_set','bed_set','facilities','hotel__tags','hotel__services').filter(hotel_id=hotel_id).order_by('id')
```
가독성을 위해 코드 줄 분리해주세요

ecila7290 replied:
수정했습니다.

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
+        rooms=Room.objects.select_related('hotel').prefetch_related('roomimage_set','bed_set','facilities','hotel__tags','hotel__services').filter(hotel_id=hotel_id).order_by('id')
+        detail=[{'common_info':{
+              'hotel_image_url'   : rooms.first().hotel.thumbnail_url,
+              'room_count'        : len(rooms),
```
길이를 재는 django 메소드도 한 번 사용해보세용!

ecila7290 replied:
count()메소드 사용하는 것으로 수정했습니다!

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
+        detail=[{'common_info':{
+              'hotel_image_url'   : rooms.first().hotel.thumbnail_url,
+              'room_count'        : len(rooms),
+              'hotel_name'        : rooms.first().hotel.name,
+              'hotel_english_name': rooms.first().hotel.english_name,
+              'hotel_introduction': rooms.first().hotel.introduction,
```
good !

ecila7290 replied:
👍

soheon-lee commented on hotel/views.py:
```diff
@@ -3,29 +3,122 @@
+                'min_people'       : room.min_people,
+                'max_people'       : room.max_people,
+                'area'             : f"{int(room.area)}{chr(0x33A1)}",
+                'bed'              : [{'bed_type':bed.bed_type.name,'number_of_beds':bed.number}for bed in room.bed_set.all()],
+                'tags'             : [tag.name for tag in room.hotel.tags.all()] if len(room.hotel.tags.all())<=3 else [tag.name for tag in room.hotel.tags.all()[:3]],
+                'price'            : chr(0x20A9)+"{:,}".format(int(room.price))+'~',
+                'facility'         : [{'name':facility.name, 'icon_url':facility.icon_url} for facility in room.facilities.all()],
+                'service'          : [{'name':service.name, 'icon_url':service.icon_url} for service in room.hotel.services.all()]
```
가독성을 위해 코드 줄 분리해주세요

ecila7290 replied:
수정했습니다.

ecila7290 committed:
Edit: edited hotel/views
 - 배너, 매거진 view에서 불러오는 인스턴스 개수를 변수로 수정
 - PicksView에서 필터 셋을 딕셔너리로 수정
 - 가독성을 위한 코드 일부 수정

ecila7290 committed:
Add: added banners as query string
 - 배너로 보내는 데이터의 양을 쿼리스트링으로 받고, 쿼리스트링이 없을 경우 모든 데이터를 보내는 것으로 합의

ecila7290 committed:
Edit: edited main banner view
 - 쿼리스트링으로 받은 banners를 정수형으로 수정

ecila7290 committed:
Merge branch 'master' into feature/detail_page

ecila7290 committed:
Edit: edited login view and utils
 - jwt의 algorithms->algorithm으로 수정

ecila7290 committed:
Merge branch 'master' into feature/detail_page
 - resolved conflict

ecila7290 committed:
Fix: resolved conflict

ecila7290 committed:
Resolved conflict

soheon-lee commented:
LGTM

soheon-lee commented on hotel/views.py:
```diff
@@ -20,7 +24,8 @@ def get(self, request):
     def get(self, request):
-        hotels    = Hotel.objects.select_related('category','location').all().order_by('id')[:2]
+        MAGAZINE  = 2
+        hotels    = Hotel.objects.select_related('category','location').all().order_by('id')[:MAGAZINE]
```
```suggestion
        hotels    = Hotel.objects.select_related('category','location').order_by('id')[:MAGAZINE]
```
없어도 동일하게 동작합니다 :)

soheon-lee commented on hotel/views.py:
```diff
@@ -33,9 +38,50 @@ def get(self, request):
-        hotels=Hotel.objects.all()
-        picks=[
-            {
+        hotels=Hotel.objects.select_related('category','location').prefetch_related('room_set','tags').all().order_by('id')
+        filter_set={}
+        if request.GET.get('category'):
+            filter_set['category__name'] = request.GET['category'].strip("'")
+        if request.GET.get('location'):
+            filter_set['location__name__contains'] = request.GET['location'].strip("'")
+        
+        for key,value in filter_set.items():
+            if filter_set == {}:
+                break
+            hotels=hotels.filter(**{key:value})
```
```suggestion
       
        filter_set={}
        if request.GET.get('category'):
            filter_set['category__name'] = request.GET['category'].strip("'")
        if request.GET.get('location'):
            filter_set['location__name__contains'] = request.GET['location'].strip("'")
       
        hotels=Hotel.objects.select_related('category','location').prefetch_related('room_set','tags').filter(**filter_set).order_by('id') 

```

soheon-lee replied:
혹시 이렇게해도 같은 경우가 될까요?
